api/saml/provider.py

```diff
@@ -209,9 +211,32 @@ def _get_idp_display_names(identity_provider_index, identity_provider):
             )
             return [SAMLLocalizedMetadataItem(display_name, language="en")]
 
-    def _create_token(self, db, patron, subject, cm_session_lifetime):
-        """Creates a Credential object that ties the given patron to the
-        given provider token.
+    @staticmethod
+    def _create_token_value(subject):
+        """Serialize the SAML subject.
+
+        :param subject: SAML subject
+        :type subject: api.saml.metadata.model.SAMLSubject
+
+        :return: Token value
+        :rtype: str
+        """
+        subject = deepcopy(subject)
+
+        # We should not save a transient Name ID because it changes each time
+        if (
+            subject.name_id
+            and subject.name_id.name_format == SAMLNameIDFormat.TRANSIENT.value
+        ):
+            subject.name_id = None
```

Comment:
I assume the reason this code isn't in `SAMLSubjectJSONEncoder` is that not only do we not want to put a transient name into the token, we don't even want to store it in the `Subject` object.  Is that right

Reply:
Actually, I wanted to keep the `SAMLSubject` object untouched, I didn't want to mutate the state during saving process because I thought it might lead to some unwanted side-effects.

Reply:
That makes sense, but doesn't `subject.name_id = None` change the state of the `SAMLSubject`?

Reply:
Yes, it does but it changes the deep copy of it so the original object left unchanged.

Reply:
Got it, thanks.